js/setup.js

```diff
@@ -46,6 +47,100 @@ chrome.devtools.network.onRequestFinished.addListener(
     requestRenderer.addNetworkRequestEntry(response.request.url);
 });
 
+/**
+ * This function will be injected into the main page and be executed there.
+ * @param {string} extensionId the chrome.runtime.id of the Mp4Inspector
+ */
+function overWriteSourceBufferAppendData(extensionId) {
+  const originalAppendBuffer = SourceBuffer.prototype.appendBuffer;
+  console.log('injecting append buffer listener');
+
+  SourceBuffer.prototype.appendBuffer = function (data) {
+    originalAppendBuffer.call(this, data);
+
+    try {
+      if (!extensionId) {
+        return;
+      }
+
+      chrome.runtime.sendMessage(extensionId, {
+        type: 'segment-appended',
+        data: arrayBufferToBase64(data),
+      });
+    } catch (err) {
+      console.warn(err);
+    }
+  };
+  function arrayBufferToBase64(arrayBuffer) {
+    let base64String = '';
+    const data = new Uint8Array(arrayBuffer);
+
+    for (let idx = 0; idx < data.length; idx++) {
+      base64String += String.fromCharCode(data[idx]);
+    }
+
+    return window.btoa(base64String);
+  }
+}
+
+function getCurrentTab() {
+  return new Promise((resolve, reject) => {
+    function onTabsAvailable(tabArray) {
+      if (tabArray && tabArray.length > 0) {
+        resolve(tabArray[0]);
+      } else {
+        reject('No active tab');
+      }
+    }
+    chrome.tabs.query({ currentWindow: true, active: true }, onTabsAvailable);
+  });
+}
+
+function ensureScriptingPermission(activeTab) {
+  return new Promise((resolve, reject) => {
+    if (!activeTab.url) {
+      // it seems that the url is only exposed when the url is listed in the
+      // host_permissions or externally_connectable entries in the manifest.json
+      // if this is not filled, this is a good indicator for missing permissions
+      reject(
+        'activeTab.url is missing, so we cant check for permissions on the current page'
+      );
+    }
+
+    const permissions = {
+      permissions: ['scripting'],
+      origins: [activeTab.url],
+    };
+    chrome.permissions.contains(permissions, function (hasPermission) {
+      if (hasPermission) {
+        resolve(activeTab);
+      } else {
+        reject('No scripting permission on the url: ' + activeTab.url);
+      }
+    });
+  });
+}
+
+function installSourceBufferOverwrite() {
+  getCurrentTab()
+    .then((activeTab) => ensureScriptingPermission(activeTab))
```

Comment:
Can be simplified: `.then(ensureScriptingPermission)`

Reply:
oh right, i knew there should be an easier way :D